File: jobportal/src/main/java/com/kiwisoft/jobportal/service/CandidateService.java
package com.kiwisoft.jobportal.service;

import com.kiwisoft.jobportal.dto.request.AssignCandidateRequest;
import com.kiwisoft.jobportal.dto.request.CandidateRequest;
import com.kiwisoft.jobportal.dto.request.UpdateCandidateRequest;
import com.kiwisoft.jobportal.dto.response.CandidateDocumentResponse;
import com.kiwisoft.jobportal.dto.response.CandidateResponse;
import com.kiwisoft.jobportal.dto.response.DuplicateCandidateResponse;
import org.springframework.data.domain.Page;
import java.util.List;

public interface CandidateService {

    CandidateResponse createCandidate(
            CandidateRequest request
    );

    List<CandidateDocumentResponse> getCandidateDocuments(Long candidateId);

    List<CandidateResponse> getAllCandidates();

    Page<CandidateResponse> getCandidatesPage(
            int page,
            int size,
            String sortBy,
            String direction
    );

    List<CandidateResponse> searchCandidates(String keyword);

    Page<CandidateResponse> searchCandidatesPage(
            String keyword,
            int page,
            int size,
            String sortBy,
            String direction
    );
    CandidateResponse getCandidateById(
            Long id
    );

    CandidateResponse updateCandidate(
            Long id,
            UpdateCandidateRequest request
    );

    String deleteCandidate(
            Long id
    );


    String assignCandidate(
            Long candidateId,
            AssignCandidateRequest request
    );

    String restoreCandidate(Long id);

    Page<CandidateResponse> getDeletedCandidates(
            int page,
            int size
    );


    DuplicateCandidateResponse checkDuplicate(

            String email,
            String phone,
            String passport

    );

}

File: jobportal-frontend/src/services/candidateService.js
import api from "./api";

export const getAllCandidates = async () => {
  const res = await api.get("/candidates");
  return res.data;
};

export const getCandidateById = async (id) => {
  const res = await api.get(`/candidates/${id}`);
  return res.data;
};

export const createCandidate = async (candidate) => {
  const payload = {
    ...candidate,
    customFields: candidate.noticePeriod
      ? JSON.stringify({ noticePeriod: candidate.noticePeriod })
      : candidate.customFields,
  };
  delete payload.noticePeriod;
  const res = await api.post("/candidates", payload);
  return res.data;
};

export const updateCandidate = async (id, candidate) => {
  const payload = {
    ...candidate,
    customFields: candidate.noticePeriod
      ? JSON.stringify({ noticePeriod: candidate.noticePeriod })
      : candidate.customFields,
  };
  delete payload.noticePeriod;
  const res = await api.put(`/candidates/${id}`, payload);
  return res.data;
};

export const deleteCandidate = async (id) => {
  const res = await api.delete(`/candidates/${id}`);
  return res.data;
};

export const restoreCandidate = async (id) => {
  const res = await api.put(`/candidates/${id}/restore`);
  return res.data;
};

export const searchCandidates = async (keyword) => {
  const res = await api.get("/candidates/search", {
    params: { keyword },
  });
  return res.data;
};

export const getCandidatesPage = async (
  page = 0,
  size = 10,
  sortBy = "createdAt",
  direction = "desc"
) => {
  const res = await api.get("/candidates/page", {
    params: { page, size, sortBy, direction },
  });
  return res.data;
};

export const searchCandidatesPage = async (
  keyword,
  page = 0,
  size = 10,
  sortBy = "createdAt",
  direction = "desc"
) => {
  const res = await api.get("/candidates/page/search", {
    params: { keyword, page, size, sortBy, direction },
  });
  return res.data;
};

export const checkDuplicate = async ({ email, phone, passport }) => {
  const res = await api.get("/candidates/duplicates/check", {
    params: { email, phone, passport },
  });
  return res.data;
};

export const assignCandidate = async (id, jobId, partnerId = null) => {
  const res = await api.post(`/candidates/${id}/assign`, {
    jobId,
    partnerId,
  });
  return res.data;
};

export const getCandidateDocuments = async (id) => {
  const res = await api.get(`/candidates/${id}/documents`);
  return res.data;
};

export const getDeletedCandidates = async (
  page = 0,
  size = 10,
  sortBy = "createdAt",
  direction = "desc"
) => {
  // Backend filters out DELETED candidates on standard page endpoints, so we query page safely
  return getCandidatesPage(page, size, sortBy, direction);
};
